# Visualization of Syntgeny
This code is for visualization of [Syntgeny](https://github.com/BIONF/Syntgeny/tree/main), which detects the conservation of genes' order over a set of taxa.

# Before use
Before using it, we need to have two files, `nodes.dmp` and `names.dmp`, which are used to recover the full name of the species.
They can be downloaded from NCBI, just click [here](https://ftp.ncbi.nih.gov/pub/taxonomy/) to get the zip file `taxdump.tar.gz` and put the two files we need into the same folder as the code file.

Additionally, the packages we need are listed in `requirements.txt`, they can be installed using the following command.  
ps. Maybe you already have `kaleido`, but with the most recent version you may not be able to run the code successfully, it is recommended to use an older version.
```
pip install -r requirements.txt
```

# Usage
Two parameters are needed, `tsv file` generated by Syntgeny and the output `format`.
Supported format parameters include html, png, jpg, jpeg, webp, svg, pdf.
The default format is html, when no format is given.

Example command
```
python syntgeny_v.py test.syntgeny.tsv png
```

# Example Output
You can try it, use the example command for the provided tsv file, and you will see the following picture.  

Species and contigs are shown on the left, the red lines at the beginning and end represent the boundaries of the contigs, and each distance is represented by the number of genes and shown in the box.   

Each colored-filled arrow represents a gene, orthologous genes share the same color, and the direction of the arrow represents the orientation of the gene, i.e. an arrow to the right means that the gene is located in the positive strand, and an arrow to the left means the negative strand.
![demo](/test.syntgeny.png)


